output

```diff
@@ -0,0 +1,137 @@
+#include "swerve/Vision.h"
+#include "swerve/Drivetrain.h"
+#include <frc/DataLogManager.h>
+
+Vision::Vision(std::function<units::degree_t()> get_angle_fn)
+    : get_angle{get_angle_fn} {};
+
+std::vector<std::optional<frc::Pose2d>> Vision::get_bot_position()
+{
+  frc::SmartDashboard::PutNumber("vision step", 1);
+  std::vector<std::optional<frc::Pose2d>> ret;
+  // for (auto &i : m_photoncam_vec)
+  // {
+  //   frc::SmartDashboard::PutNumber("vision step", 2);
+  //   auto result = i.camera->GetLatestResult();
+  //   frc::SmartDashboard::PutNumber("vision step", 2.5);
+  //   frc::SmartDashboard::PutNumber("vision is present", result.HasTargets());
+  //   if (result.MultiTagResult().result.isPresent)
+  //   {
+
+  //     frc::SmartDashboard::PutNumber("vision step", 3);
+  //     auto pose = i.multitag_estimator.Update(result);
+  //     if (pose)
+  //     {
+
+  //       frc::SmartDashboard::PutNumber("vision step", 4);
+  //       frc::SmartDashboard::PutNumber("pv/x", pose.value().estimatedPose.X().value());
+  //       frc::SmartDashboard::PutNumber("pv/y", pose.value().estimatedPose.Y().value());
+  //       frc::SmartDashboard::PutNumber("pv/get_angle()", get_angle().value());
+  //       // if (CONSTANTS::IN_THRESHOLD<units::degree_t>(
+  //       //         pose.value().estimatedPose.Rotation().ToRotation2d().Degrees(),
+  //       //         get_angle(), 3_deg))
+  //       // {
+  //       ret.push_back(pose.value().estimatedPose.ToPose2d());
+  //       // }
+  //     }
+  //   }
+  //   else if (result.HasTargets())
+  //   {
+  //     auto pose = i.singletag_estimator.Update(result);
+  //     ret.push_back(pose.value().estimatedPose.ToPose2d());
+  //   }
+  // }
+
+  for (auto &i : m_limelight_vec)
+  {
+    if ((i->GetNumber("pipeline", -1) == 0) &&
+        (i->GetNumber("tv", 0) >= 0.5))
+    {
+      auto results = i->GetNumberArray("botpose_wpiblue",
+                                       std::vector<double>(6));
+
+      ret.push_back(frc::Pose2d(units::meter_t{results[0]},
+                                units::meter_t{results[1]},
+                                frc::Rotation2d(units::degree_t{results[5]})));
+    }
+  }
+
+  std::vector<double> printvec_x;
+  std::vector<double> printvec_y;
+  for (auto &i : ret)
+  {
+    if (i)
+    {
+      printvec_x.push_back(i.value().X().value());
+      printvec_y.push_back(i.value().Y().value());
+    }
+  }
+  frc::SmartDashboard::PutNumberArray("pv/x arr", printvec_x);
+  frc::SmartDashboard::PutNumberArray("pv/y arr", printvec_y);
+  return ret;
+}
+
+std::optional<units::degree_t> Vision::get_coral_angle()
```

Comment:
Is this needed? LL4 is not compatible with Coral.

Reply:
The neural network pipeline API is the same between LL2+ and LL4. The code doesn't need to change.